File: LoanMe/customer.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

/**
 * @author Divyansh
 *
 */
public class customer extends Thread{
	String customer_name;
	ArrayList<bank> potential_lenders;
	Random r = new Random();
	int random_amount, amount, actual_amount;
	bank random_bank; 
	
	public customer(String customer_name, int amount, ArrayList<bank> bank_list) {
		this.customer_name = customer_name;
		this.amount = amount;
		this.potential_lenders = new ArrayList<>(bank_list);
		this.actual_amount = amount;
		start();
	}
	
	public void run() {
		do {
				random_amount = r.nextInt(50);
				if(potential_lenders.size()>1) {
					int random = r.nextInt(potential_lenders.size());
					random_bank = potential_lenders.get(random);
				}
				else if(potential_lenders.size()==1){
					random_bank = potential_lenders.get(0);
				}
				else {
					break;
				}
			if(random_amount<=amount && random_amount>0) {
				int low = 10;
				int high = 100;
				int result = r.nextInt(high-low) + low;
				
				try {
					Thread.sleep(result);
				} catch (Exception e) {
					// TODO: handle exception
				}
				String response = random_bank.request(random_amount);
				System.out.println(customer_name+" requests a loan of "+random_amount+" dollar(s) from "+random_bank.bank_name);
				
				if(response=="approve") {
					amount = amount - random_amount;
					System.out.println(random_bank.bank_name+" approves a loan of "+random_amount+" dollar(s) from "+customer_name);
					
					if(amount==0) {
						for(int i=0;i<potential_lenders.size();i++) {
							potential_lenders.get(i).stopThread();
						}
						break;
					}
				}
				else if(response=="deny") {
					random_bank.stopThread();
					potential_lenders.remove(random_bank);
					System.out.println(random_bank.bank_name+" denies a loan of "+random_amount+" dollar(s) from "+customer_name);
				}
				
			}
		} while (amount!=0 || potential_lenders.size()!=0);
		if(amount==0) {
			System.out.println(customer_name+" has reached the objective of "+actual_amount+" dollar(s). Woo Hoo!");
		}
		else if(amount>0){
			System.out.println(customer_name+" was only able to borrow "+(actual_amount-amount)+" dollar(s). Boo Hoo!");
		}
		
	}
	
}

File: LoanMe/customer.erl
%%%-------------------------------------------------------------------
%%% @author Divyansh
%%% @copyright (C) 2019, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 18. Jun 2019 21:34
%%%-------------------------------------------------------------------
-module(customer).
-author("Divyansh").

%% API
-export([childCustomer/4]).

childCustomer(Customer_name, Amount, Bank_list, Actual_amount) ->
  if
    Amount /= 0 ->
      if
        length(Bank_list) /= 0 ->
         Random_amount =  if
                            Amount < 50 ->
                              rand:uniform(Amount);
                            true ->
                              rand:uniform(50)
                          end,

          Random_bank = lists:nth(rand:uniform(length(Bank_list)), Bank_list),
          if
            Random_amount /= 0 ->
              Low = 10,
              High = 100,
              Random_sleep = rand:uniform(High - Low) + Low,
              sleep(Random_sleep),
              io:format("~p requests a loan of ~p dollar(s) from ~p ~n ", [Customer_name, Random_amount, Random_bank]),
              Pid = whereis(Random_bank),
              Pid ! {self(), Random_amount};
            true ->
              true
          end,

          receive
            {approve, Random_amount} ->
              io:format("~p approves a loan of ~p dollar(s) from ~p ~n ", [Random_bank, Random_amount, Customer_name]),
              childCustomer(Customer_name, Amount - Random_amount, Bank_list, Actual_amount);

            {deny} ->
              io:format("~p denies a loan of ~p dollar(s) from ~p ~n ", [Random_bank, Random_amount, Customer_name]),
              childCustomer(Customer_name, Amount, lists:filter(fun(Elem) -> not lists:member(Elem, [Random_bank]) end, Bank_list), Actual_amount)
          end;

        true ->
          if
            Amount == 0 ->
              io:format("~p has reached the objective of ~p dollar(s). Woo Hoo!~n ", [Customer_name, Actual_amount]);
            Amount > 0 ->
              io:format("~p was only able to borrow ~p dollar(s). Boo Hoo! ~n ", [Customer_name, Actual_amount - Amount]);
            true ->
              true
          end
      end;
    true ->
      if
        Amount == 0 ->
          io:format("~p has reached the objective of ~p dollar(s). Woo Hoo!~n ", [Customer_name, Actual_amount]);
        Amount > 0 ->
          io:format("~p was only able to borrow ~p dollar(s). Boo Hoo! ~n ", [Customer_name, Actual_amount - Amount]);
        true ->
          true
      end
  end.

sleep(T) ->
  receive
  after T ->
    true
  end.